### PR TITLE
Data_oM: Adding Dataset and Source object

### LINKED ISSUE #576: Data_oM: Add Dataset object
#### Description:

Need a `Dataset` class to work with all our serialised Datasets used by the Library_Engine.

Object to have properties:

- Source (this might need to be a separate object)
- Date of serialisation
- Data

Question is if this is best put in the Data_oM or if we should add a Library_oM for this. Important to not get this mixed up with any of the work done by @epignatelli 

@al-fisher @adecler @rwemay @FraserGreenroyd @JosefTaylor @enarhi @kayleighhoude @peterjamesnugent  any input welcome.

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### Issues addressed by this PR

Closes #576 

Adding Dataset object to be used with the serialised data in Datasets folder together with library engine. Added to formalise storing metadata such as time of creating and various soruce information

### Additional comments

@peterjamesnugent @kayleighhoude @JosefTaylor @enarhi @FraserGreenroyd

### CODE REVIEW

**Comment by adecler** [comment]
Looks good to me. 
I just wonder if `HyperLink` and `FilePath` shouldn't be collapsed into one property.

**Comment by al-fisher** [comment]
Thanks @IsakNaslundBh 
This is looking good. In addition to merging the `SourceLink` - wonder if we should do the same for 
`FigureReference`, `CodeReference` and `TableReference` - would all three be used simultaneously? or can we simply combine as a single field. 

Clarity could also be placed on each field by adding a `DiscriptionAttribute`

**Comment by IsakNaslundBh** [comment]
Added first-pass descriptions to most of the properties. Any comments welcome.

Also not to sure of the Location name for country/region. If anyone has a better suggestion for that, it is more than welcome

File: Data_oM/Library/Source.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+        public string Schema { get; set; } = "";
+
+        [Description("Language of the source information")]
+        public string Language { get; set; } = "";
+
+        [Description("Country/Region of the source information")]
+        public string Location { get; set; } = "";
+
+        [Description("Any applicable copyright information assosiated with the data")]
```
Comment (peterjamesnugent) [inline Data_oM/Library/Source.cs]: For consistency we should always refer to it as source information and not use variations of data, information etc.

File: Data_oM/Library/Source.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        [Description("Title of the source. For example Book title, Paper title, Code title etc.")]
+        public string Title { get; set; } = "";
+
+        [Description("Author of the source of the information")]
+        public string Author { get; set; } = "";
+
+        [Description("Table, figure or code reference etc.")]
+        public string ItemReference { get; set; } = "";
+
+        [Description("Version of the source used")]
```
Comment (peterjamesnugent) [inline Data_oM/Library/Source.cs]: Version of the source information used

File: Data_oM/Library/Source.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Description("Hyperlink or file path to where the source information used can be found")]
+        public string SourceLink { get; set; } = "";
+
+        [Description("Title of the source. For example Book title, Paper title, Code title etc.")]
+        public string Title { get; set; } = "";
+
+        [Description("Author of the source of the information")]
```
Comment (peterjamesnugent) [inline Data_oM/Library/Source.cs]: Author of the source information

File: Data_oM/Library/Source.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        [Description("Table, figure or code reference etc.")]
+        public string ItemReference { get; set; } = "";
+
+        [Description("Version of the source used")]
+        public string Version { get; set; } = "";
+
+        [Description("Publisher of the source information")]
+        public string Publisher { get; set; } = "";
+
+        public string Schema { get; set; } = "";
```
Comment (peterjamesnugent) [inline Data_oM/Library/Source.cs]: There's no description here, are we expecting it to be a scheme reference such as name?

**Pull request merged**